hphp/runtime/vm/jit/vasm-arm.cpp

```diff
@@ -1197,45 +1232,6 @@ Y(storew, m)
 
 #undef Y
 
-#define ISAR vixl::Assembler::IsImmArithmetic(value)
-#define ISLG(w) vixl::Assembler::IsImmLogical(value, w)
-#define U2(v0, v1) v0, v1
-
-#define Y(vasm_opc, lower_opc, load_opc, chk, imm, use)     \
-void lower(const VLS& e, vasm_opc& i, Vlabel b, size_t z) { \
-  lower_impl(e.unit, b, z, [&] (Vout& v) {                  \
-    auto value = safe_cast<int64_t>(i.s0.q());              \
-    if (chk) {                                              \
-      v << i;                                               \
-    } else {                                                \
-      auto s0 = v.makeReg();                                \
-      v << load_opc{imm, s0};                               \
-      v << lower_opc{s0, use, i.sf};                        \
-    }                                                       \
-  });                                                       \
-}
-
-Y(addli, addl, ldimml, ISAR, i.s0, U2(i.s1, i.d))
-Y(addqi, addq, ldimmq, ISAR, Immed64(value), U2(i.s1, i.d))
-Y(andbi, andb, ldimmb, ISLG(32), i.s0, U2(i.s1, i.d))
-Y(andli, andl, ldimml, ISLG(32), i.s0, U2(i.s1, i.d))
-Y(andqi, andq, ldimmq, ISLG(64), Immed64(value), U2(i.s1, i.d))
-Y(cmpli, cmpl, ldimml, ISAR, i.s0, i.s1)
-Y(cmpqi, cmpq, ldimmq, ISAR, Immed64(value), i.s1)
-Y(orqi, orq, ldimmq, ISLG(64), Immed64(value), U2(i.s1, i.d))
-Y(subli, subl, ldimml, ISAR, i.s0, U2(i.s1, i.d))
-Y(subqi, subq, ldimmq, ISAR, Immed64(value), U2(i.s1, i.d))
```

Comment:
My guess: There are different ways to handle the fact that not all vasm instructions "are existing" on aarch64. E.g. `cmpb{}` does not exist. So one way is to lower it to `cmpl{}` and have no emit function for `cmpb{}`. That is also true for all these instructions (e.g. splitting up cmpqi{} if the immediate is to big). However, this approach (lower everything what cannot be directly emitted) is not used throughout the whole code. Once I gave it a try to enforce this at least for `ldimmN{}` (see https://github.com/cmuellner/hhvm/commit/4ba6b5756555852bfba45a2dec1638e2f3d4bf06), but that ended up in having to touch the register allocator, which assumes `ldimmb{}` to be ok to emit (the register allocator is running after lowerForArm()).

In general I like the idea of lowering everything what cannot be translated to aarch64 directly, but I am not sure if it is worth it given that a) vixl can handle that anyways (we would have to reimplement vixl-features with the risk of making it worse) and b) non-arm-specific code might need to be aware that it cannot use some vasm instructions (e.g. the register allocator). Therefore I prefer to only lower vasm instructions, which can be lowered or need to be lowered (for some reason). But I am open here to change my mind. :)